### PR TITLE
implement new rule to disallow test import in production code

### LINKED ISSUE #60: No import of test/spec files from production code
We need a rule similar to this: https://github.com/avajs/eslint-plugin-ava/blob/main/docs/rules/no-import-test-files.md

No `.test` or `.spec` module should be imported from production (i.e. non-test, non-spec file) code.

### PR DESCRIPTION
**Pull request by le-cong**
Fixes #60

### CODE REVIEW

File: src/no-test-import-in-production-code.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      description:
+        'Importing test files in production code is not allowed since it can lead to problems once the test files are removed from the build',
+      url: 'https://github.com/checkdigit/eslint-plugin',
+    },
+    messages: {
+      [NO_TEST_IMPORT_IN_PRODUCTION_CODE]: `Importing test files in production code is not allowed`,
+    },
+  },
+  create(context) {
+    const testFilePattern = /\.(?<testFileSuffix>test|spec)(?<tsFileSuffix>\.ts)?$/u;
```
Comment (carlansley) [inline src/no-test-import-in-production-code.ts:26]: do we need this to be just `.ts` files?  i think we could loosen the regex, so you can't import test/spec .js files or .json files etc.

File: src/no-test-import-in-production-code.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    return {
+      ImportDeclaration(node) {
+        const filename = context.filename;
+
+        // skip test files, only production code should be checked
+        if (filename.match(testFilePattern)) {
```
Comment (carlansley) [inline src/no-test-import-in-production-code.ts:33]: this kind of thing is normally done in the config.  maybe the rule should be "no-test-import", and we only have it turned on for prod code (and disabled for test/spec code) in eslint-config?

**Review by carlansley — CHANGES_REQUESTED** [review]

**Review by carlansley — CHANGES_REQUESTED** [review]

File: src/no-test-import.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+export interface NoTestImportRuleOptions {
+  testFilePattern?: string;
+}
+export const NO_TEST_IMPORT = 'NO_TEST_IMPORT';
+const DEFAULT_TEST_FILE_PATTERN = '\\.(test|spec)(\\.\\w+)?$';
+
+export default {
+  meta: {
```
Comment (carlansley) [inline src/no-test-import.ts:22]: i think you can/should set the default options here?  `defaultOptions` or something?  (vs the code below)

**Review by carlansley — CHANGES_REQUESTED** [review]

File: src/no-test-import.ts (hunk 3 above)
Comment (le-cong) [reply src/no-test-import.ts:22]: i could be wrong, by checking the [document](https://eslint.org/docs/latest/extend/custom-rules) it doesn't seem to be a way to set defaultOptions. i did a minor adjustment and added a schema for options validation purpose.

**Review by carlansley — APPROVED** [review]

**Pull request merged**